src/wallet/wallet2.cpp

```diff
@@ -3097,6 +3194,10 @@ bool wallet2::load_keys(const std::string& keys_file_name, const epee::wipeable_
   if(!m_watch_only && !m_multisig)
     r = r && hwdev.verify_keys(keys.m_spend_secret_key, keys.m_account_address.m_spend_public_key);
   THROW_WALLET_EXCEPTION_IF(!r, error::invalid_password);
+
+  if (r)
+    setup_keys(password);
```

Comment:
The call to `store_keys` on L3183 should come after this line, because `store_keys` assumes that the secret keys are encrypted when it's called.
```diff
@@ -3177,16 +3177,6 @@ bool wallet2::load_keys(const std::string& keys_file_name, const epee::wipeable_
     {
       m_account.decrypt_keys(key);
     }
-    else
-    {
-      // rewrite with encrypted keys, ignore errors
-      bool saved_ret = store_keys(keys_file_name, password, m_watch_only);
-      if (!saved_ret)
-      {
-        // just moan a bit, but not fatal
-        MERROR("Error saving keys file with encrypted keys, not fatal");
-      }
-    }
   }
   const cryptonote::account_keys& keys = m_account.get_keys();
   hw::device &hwdev = m_account.get_device();
@@ -3196,7 +3186,19 @@ bool wallet2::load_keys(const std::string& keys_file_name, const epee::wipeable_
   THROW_WALLET_EXCEPTION_IF(!r, error::invalid_password);
 
   if (r)
+  {
     setup_keys(password);
+    if (!encrypted_secret_keys)
+    {
+      // rewrite with encrypted keys, ignore errors
+      bool saved_ret = store_keys(keys_file_name, password, m_watch_only);
+      if (!saved_ret)
+      {
+        // just moan a bit, but not fatal
+        MERROR("Error saving keys file with encrypted keys, not fatal");
+      }
+    }
+  }
 
   return true;
 }
```

Reply:
Also I realized that `m_keys_file_locker.reset();` should be called after `store_keys` because it internally locks the keys file. Otherwise, an error message will be shown when opening a legacy wallet file (but it seems harmless):

> ERROR	net.http	src/common/util.cpp:234	Failed to lock tests/data/wallet_9svHk1.keys: Resource temporarily unavailable

Reply:
Ah nevermind, this part is from a patch that's needed, you can ignore this, it's only been fixed in the PR this patch is for.

Reply:
In fact, I think I might just bundle all the secret key related patches in a single PR, because this is getting really annoying having to cherry-pick needed patches in order for the PR to work.

Reply:
> I think I might just bundle all the secret key related patches in a single PR
+1

I found the portable wallet serialization unit test failing with the current PR, and came up with the above patch to make it pass the test. Please make sure that the test passes in any case.